### PR TITLE
feat: add a cloud storage write precondition

### LINKED ISSUE #259: Upload images with an ifGenerationMatch 0 parameter
To avoid redundant, concurrent writes to the image cloud storage bucket (which can trigger rate limit errors), we should be able to limit the writes by specifying a precondition to match the first generation of the object.

So once the first write to the object completes, subsequent attempts should fail the precondition.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/CONSVC-1205)

### PR DESCRIPTION
**Pull request by pjenvey**
## Description

so we only issue writes when the object doesn't already exist, avoiding rate
limiting errors

This supersedes the bits of #258 adding the precondition check

## Testing

Please run the img_storage tests with a test bucket setup

## Issue(s)

Closes #259

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -26,7 +26,7 @@ bytes = "1.0"
 cadence = "0.26"
 chrono = "0.4"
 docopt = "1.1"
-cloud-storage = "0.6" # 0.7+ includes request 0.11, tokio 1.4
+cloud-storage = { git = "https://github.com/mozilla-services/cloud-storage-rs", branch = "feat/create_with_params-0.6.2" } # 0.7+ includes request 0.11, tokio 1.4
```
Comment (pjenvey) [inline Cargo.toml:29]: this branch could use a separate review, I can set up a PR to merge it to another branch for that sake

File: src/server/img_storage.rs [hunk 2]
```diff
@@ -121,7 +121,7 @@ impl Default for StoreImage {
 #[derive(Debug)]
 pub struct StoredImage {
     pub url: uri::Uri,
-    pub object: Object,
+    pub object: Option<Object>,
```
Comment (jrconlin) [inline src/server/img_storage.rs:124]: we can probably drop this. It's not really used.

**Review by jrconlin — APPROVED** [review]

**Comment by jrconlin** [comment]
Once this lands, really tempted to redo #258 to drop the useless vendor and just be "use redis as a shared image metadata cache". It'll still be a draft we can keep in our collective back pocket, but i'll note the redis integration was fairly easy.

**Review by jrconlin — APPROVED** [review]

File: src/server/img_storage.rs (hunk 2 above)
Comment (pjenvey) [reply src/server/img_storage.rs:124]: woops, almost forgot this

Comment (pjenvey) [reply src/server/img_storage.rs:124]: And then I actually did forget this (forgot to push it)

**Pull request merged**